<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Промежуточные-выводы:" data-toc-modified-id="Промежуточные-выводы:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Промежуточные выводы:</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Сводная-таблица-результатов" data-toc-modified-id="Сводная-таблица-результатов-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Сводная таблица результатов</a></span></li><li><span><a href="#Промежуточные-выводы" data-toc-modified-id="Промежуточные-выводы-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Промежуточные выводы</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Взвешивание классов</a></span></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Увеличение выборки</a></span></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Уменьшение выборки</a></span></li><li><span><a href="#Промежуточные-выводы:" data-toc-modified-id="Промежуточные-выводы:-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Промежуточные выводы:</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Промежуточные-выводы" data-toc-modified-id="Промежуточные-выводы-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Промежуточные выводы</a></span></li></ul></li><li><span><a href="#Итоговые-выводы:" data-toc-modified-id="Итоговые-выводы:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Итоговые выводы:</a></span><ul class="toc-item"><li><span><a href="#Ход-работы:" data-toc-modified-id="Ход-работы:-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Ход работы:</a></span></li><li><span><a href="#Результаты:" data-toc-modified-id="Результаты:-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Результаты:</a></span></li></ul></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

Цель проекта: найти модель со значением F1-меры более 0.59

Необходимые этапы для достижения цели:
1. Подготовка данных
2. Построение моделей, не учитывая дисбаланс классов
3. Борьба с дисбалансом
4. Сравнение полученных результатов
5. Выбор наилучшей модели
6. Тестирование наилучшей модели

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score 
from sklearn.utils import shuffle
from sklearn.preprocessing import OneHotEncoder
import warnings
warnings.filterwarnings('ignore')

In [2]:
pip install scikit-learn==1.1.3

Note: you may need to restart the kernel to use updated packages.


In [3]:
data = pd.read_csv('Churn.csv', index_col = 'RowNumber')
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 1 to 10000
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CustomerId       10000 non-null  int64  
 1   Surname          10000 non-null  object 
 2   CreditScore      10000 non-null  int64  
 3   Geography        10000 non-null  object 
 4   Gender           10000 non-null  object 
 5   Age              10000 non-null  int64  
 6   Tenure           9091 non-null   float64
 7   Balance          10000 non-null  float64
 8   NumOfProducts    10000 non-null  int64  
 9   HasCrCard        10000 non-null  int64  
 10  IsActiveMember   10000 non-null  int64  
 11  EstimatedSalary  10000 non-null  float64
 12  Exited           10000 non-null  int64  
dtypes: float64(3), int64(7), object(3)
memory usage: 1.1+ MB


In [4]:
data.head()

CustomerId   Surname  CreditScore Geography  Gender  Age  Tenure  \
RowNumber                                                                     
1            15634602  Hargrave          619    France  Female   42     2.0   
2            15647311      Hill          608     Spain  Female   41     1.0   
3            15619304      Onio          502    France  Female   42     8.0   
4            15701354      Boni          699    France  Female   39     1.0   
5            15737888  Mitchell          850     Spain  Female   43     2.0   

             Balance  NumOfProducts  HasCrCard  IsActiveMember  \
RowNumber                                                        
1               0.00              1          1               1   
2           83807.86              1          0               1   
3          159660.80              3          1               0   
4               0.00              2          0               0   
5          125510.82              1          1               1   

           EstimatedSalary  Exited  
RowNumber                           
1                101348.88       1  
2                112542.58       0  
3                113931.57       1  
4                 93826.63       0  
5                 79084.10       0

In [5]:
data = data.drop(['CustomerId', 'Surname'], axis=1)

In [6]:
data = data.dropna()
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 1 to 9999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      9091 non-null   int64  
 1   Geography        9091 non-null   object 
 2   Gender           9091 non-null   object 
 3   Age              9091 non-null   int64  
 4   Tenure           9091 non-null   float64
 5   Balance          9091 non-null   float64
 6   NumOfProducts    9091 non-null   int64  
 7   HasCrCard        9091 non-null   int64  
 8   IsActiveMember   9091 non-null   int64  
 9   EstimatedSalary  9091 non-null   float64
 10  Exited           9091 non-null   int64  
dtypes: float64(3), int64(6), object(2)
memory usage: 852.3+ KB


Удалим столбцы, которые хранят признаки не влияющие на уход клиента из банка.

Так как в столбце tenure есть пропушенные значения, и их немного, удалим строки с пропусками.

In [9]:
features = data.drop('Exited', axis=1)
target = data['Exited']
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.4, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(
    features_valid, target_valid, test_size=0.5, random_state=12345)
print(features_train.shape)
print(features_valid.shape)
print(features_test.shape)

(5454, 10)
(1818, 10)
(1819, 10)


In [10]:
enc = OneHotEncoder(drop='first', sparse=False)
category = ['Geography', 'Gender']
enc.fit(features_train[category])
# features_train[enc.get_feature_names_out()] = enc.transform(features_train[category])
features_train[enc.get_feature_names()] = enc.transform(features_train[category])
features_train = features_train.drop(category, axis=1)
features_train.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
RowNumber                                                                  
9345               727   28     2.0  110997.76              1          1   
3797               537   26     7.0  106397.75              1          0   
7463               610   40     9.0       0.00              1          1   
1509               576   36     6.0       0.00              2          1   
4479               549   31     4.0       0.00              2          0   

           IsActiveMember  EstimatedSalary  x0_Germany  x0_Spain  x1_Male  
RowNumber                                                                  
9345                    0        101433.76         0.0       0.0      0.0  
3797                    0        103563.23         0.0       0.0      1.0  
7463                    1        149602.54         0.0       0.0      1.0  
1509                    1         48314.00         0.0       0.0      1.0  
4479                    1         25684.85         0.0       0.0      1.0

In [11]:
features_valid[enc.get_feature_names()] = enc.transform(features_valid[category])
features_valid = features_valid.drop(category, axis=1)
features_test[enc.get_feature_names()] = enc.transform(features_test[category])
features_test = features_test.drop(category, axis=1)

In [12]:
numeric = ['CreditScore', 'Tenure', 'Balance', 'Age', 'NumOfProducts', 'EstimatedSalary']

scaler = StandardScaler()
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])

### Промежуточные выводы:

1. Набор данных изучен
2. Заполнены пропуски в столбце Tenure
3. Удалены признаки, не влияющие на результат
4. Проведено кодирование категориальных признаков
5. Разделен на три выборки: обучающую, валидационную и тестовую в соотношении 3:1:1 соответственно
6. Проведено масштабирование данных

## Исследование задачи

In [13]:
result_models = []

### Решающее дерево

In [14]:
max_f1 = 0
intermediate_result = []
for depth in range(1, 20):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model.fit(features_train, target_train)  
    predicted_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    intermediate_result.append(['Дерево', depth, f1, auc_roc])
    if f1 > max_f1:
        max_f1 = f1
        max_depth = depth
        max_auc_roc = auc_roc
result_models.append(['Дерево', 'Без баланса', max_depth, max_f1, max_auc_roc, model])
intermediate_result = pd.DataFrame(data=intermediate_result, columns = ['name_models', 'depth_or_est', 'f1_score', 'auc_roc'])
print('Наилучший результат:')
print('Глубина:', max_depth)
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)
print('____________________________')
print('Промежуточные результаты:')
intermediate_result

Наилучший результат:
Глубина: 7
F1-мера:  0.5764331210191083
AUC-ROC:  0.8346049843812412
____________________________
Промежуточные результаты:


name_models  depth_or_est  f1_score   auc_roc
0       Дерево             1  0.000000  0.702719
1       Дерево             2  0.530488  0.755085
2       Дерево             3  0.372671  0.799438
3       Дерево             4  0.541455  0.821332
4       Дерево             5  0.514071  0.847103
5       Дерево             6  0.560684  0.849540
6       Дерево             7  0.576433  0.834605
7       Дерево             8  0.541329  0.806683
8       Дерево             9  0.544615  0.794063
9       Дерево            10  0.512270  0.756697
10      Дерево            11  0.512821  0.713817
11      Дерево            12  0.479433  0.700640
12      Дерево            13  0.482566  0.680500
13      Дерево            14  0.486413  0.675461
14      Дерево            15  0.474114  0.671834
15      Дерево            16  0.478203  0.674329
16      Дерево            17  0.484605  0.679703
17      Дерево            18  0.480537  0.671840
18      Дерево            19  0.486202  0.678748

### Случайный лес

In [15]:
max_f1 = 0
intermediate_result = []
for est in range(10, 200, 10):
    model = RandomForestClassifier(random_state=12345, n_estimators=est) 
    model.fit(features_train, target_train) 
    predicted_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    intermediate_result.append(['Случайный лес', est, f1, auc_roc])
    if f1 > max_f1:
        max_f1 = f1
        max_est = est
        max_auc_roc = auc_roc
result_models.append(['Случайный лес', 'Без баланса', max_est, max_f1, max_auc_roc, model])
print('Наилучший результат:')
print('Количество деревьев:', max_est)
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)
intermediate_result = pd.DataFrame(data=intermediate_result, columns = ['name_models', 'depth_or_est', 'f1_score', 'auc_roc'])
print('____________________________')
print('Промежуточные результаты:')
intermediate_result

Наилучший результат:
Количество деревьев: 130
F1-мера:  0.5917355371900826
AUC-ROC:  0.8642226423221422
____________________________
Промежуточные результаты:


name_models  depth_or_est  f1_score   auc_roc
0   Случайный лес            10  0.544538  0.834427
1   Случайный лес            20  0.569558  0.847502
2   Случайный лес            30  0.577488  0.848296
3   Случайный лес            40  0.588235  0.852204
4   Случайный лес            50  0.586885  0.855615
5   Случайный лес            60  0.584565  0.857035
6   Случайный лес            70  0.585526  0.859011
7   Случайный лес            80  0.589198  0.860991
8   Случайный лес            90  0.586093  0.860276
9   Случайный лес           100  0.585124  0.862931
10  Случайный лес           110  0.588430  0.863129
11  Случайный лес           120  0.588430  0.863392
12  Случайный лес           130  0.591736  0.864223
13  Случайный лес           140  0.587459  0.863869
14  Случайный лес           150  0.588040  0.863530
15  Случайный лес           160  0.580858  0.863610
16  Случайный лес           170  0.582781  0.863706
17  Случайный лес           180  0.584718  0.864463
18  Случайный лес           190  0.585691  0.865549

### Логистическая регрессия

In [16]:
model = LogisticRegression() 
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
f1 = f1_score(target_valid, predicted_valid)
probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
result_models.append(['Логистическиая регрессия', 'Без баланса', None, f1, auc_roc, model])
print('Результат:')
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)

Результат:
F1-мера:  0.5917355371900826
AUC-ROC:  0.8642226423221422


### Сводная таблица результатов

In [17]:
models = pd.DataFrame(data=result_models, columns = ['name_models', 'balance', 'depth_or_est', 'f1_score', 'auc_roc', 'models'])
models

name_models      balance  depth_or_est  f1_score   auc_roc  \
0                    Дерево  Без баланса           7.0  0.576433  0.834605   
1             Случайный лес  Без баланса         130.0  0.591736  0.864223   
2  Логистическиая регрессия  Без баланса           NaN  0.304000  0.773663   

                                              models  
0  DecisionTreeClassifier(max_depth=19, random_st...  
1  (DecisionTreeClassifier(max_features='auto', r...  
2                               LogisticRegression()

### Промежуточные выводы

Исследованы 4 различные модели:
- решающее дерево
- случайный лес 
- линейная регрессия

Модель случайный лес показала наилучший результат, логистическая регрессия наихудший.

## Борьба с дисбалансом

In [18]:
print('Баланс классов до разбиения на выборки')
print(target[target == 0].count())
print(target[target == 1].count())
print(target[target == 0].count() / target[target == 1].count())
print('Баланс классов у обучающей выборки')
print(target_train[target_train == 0].count())
print(target_train[target_train == 1].count())
print(target_train[target_train == 0].count() / target_train[target_train == 1].count())
print('Баланс классов у валидационной выборки')
print(target_valid[target_valid == 0].count())
print(target_valid[target_valid == 1].count())
print(target_valid[target_valid == 0].count() / target_valid[target_valid == 1].count())
print('Баланс классов у тестовой выборки')
print(target_test[target_test == 0].count())
print(target_test[target_test == 1].count())
print(target_test[target_test == 0].count() / target_test[target_test == 1].count())

Баланс классов до разбиения на выборки
7237
1854
3.9034519956850056
Баланс классов у обучающей выборки
4328
1126
3.8436944937833037
Баланс классов у валидационной выборки
1441
377
3.8222811671087533
Баланс классов у тестовой выборки
1468
351
4.182336182336182


### Взвешивание классов

In [19]:
max_f1 = 0
intermediate_result = []
for depth in range(1, 20):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth, class_weight='balanced') 
    model.fit(features_train, target_train)  
    predicted_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    intermediate_result.append(['Дерево', depth, f1, auc_roc])
    if f1 > max_f1:
        max_f1 = f1
        max_depth = depth
        max_auc_roc = auc_roc
result_models.append(['Дерево', 'Взвешивание классов', max_depth, max_f1, max_auc_roc, model])
intermediate_result = pd.DataFrame(data=intermediate_result, columns = ['name_models', 'depth_or_est', 'f1_score', 'auc_roc'])
print('Наилучший результат:')
print('Глубина:', max_depth)
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)
print('____________________________')
print('Промежуточные результаты:')
intermediate_result

Наилучший результат:
Глубина: 5
F1-мера:  0.5735449735449736
AUC-ROC:  0.8396523192522141
____________________________
Промежуточные результаты:


name_models  depth_or_est  f1_score   auc_roc
0       Дерево             1  0.506586  0.706652
1       Дерево             2  0.529762  0.758072
2       Дерево             3  0.548936  0.794927
3       Дерево             4  0.546119  0.817898
4       Дерево             5  0.573545  0.839652
5       Дерево             6  0.563089  0.832096
6       Дерево             7  0.541353  0.810604
7       Дерево             8  0.558190  0.786950
8       Дерево             9  0.505998  0.756530
9       Дерево            10  0.511013  0.730145
10      Дерево            11  0.514092  0.717877
11      Дерево            12  0.504284  0.694250
12      Дерево            13  0.480601  0.669641
13      Дерево            14  0.480204  0.671756
14      Дерево            15  0.465359  0.660770
15      Дерево            16  0.478203  0.666406
16      Дерево            17  0.467568  0.665488
17      Дерево            18  0.467914  0.665881
18      Дерево            19  0.467742  0.665043

In [ ]:
max_f1 = 0
intermediate_result = []
for est in range(10, 200, 10):
    model = RandomForestClassifier(random_state=12345, n_estimators=est, class_weight='balanced') 
    model.fit(features_train, target_train) 
    predicted_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    intermediate_result.append(['Случайный лес', est, f1, auc_roc])
    if f1 > max_f1:
        max_f1 = f1
        max_est = est
        max_auc_roc = auc_roc
result_models.append(['Случайный лес', 'Взвешивание классов', max_est, max_f1, max_auc_roc, model])
print('Наилучший результат:')
print('Количество деревьев:', max_est)
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)
intermediate_result = pd.DataFrame(data=intermediate_result, columns = ['name_models', 'depth_or_est', 'f1_score', 'auc_roc'])
print('____________________________')
print('Промежуточные результаты:')
intermediate_result

In [21]:
model = LogisticRegression(random_state = 12345, class_weight='balanced') 
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
f1 = f1_score(target_valid, predicted_valid)
probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
result_models.append(['Логистическиая регрессия', 'Взвешивание классов', None, f1, auc_roc, model])
print('Результат:')
print('F1-мера: ', f1)
print('AUC-ROC: ', auc_roc)

Результат:
F1-мера:  0.509731232622799
AUC-ROC:  0.7777792094717602


In [22]:
models = pd.DataFrame(data=result_models, columns = ['name_models', 'balance', 'depth_or_est', 'f1_score', 'auc_roc', 'models'])
models

name_models              balance  depth_or_est  f1_score  \
0                    Дерево          Без баланса           7.0  0.576433   
1             Случайный лес          Без баланса         130.0  0.591736   
2  Логистическиая регрессия          Без баланса           NaN  0.304000   
3                    Дерево  Взвешивание классов           5.0  0.573545   
4             Случайный лес  Взвешивание классов         150.0  0.581081   
5  Логистическиая регрессия  Взвешивание классов           NaN  0.509731   

    auc_roc                                             models  
0  0.834605  DecisionTreeClassifier(max_depth=19, random_st...  
1  0.864223  (DecisionTreeClassifier(max_features='sqrt', r...  
2  0.773663                               LogisticRegression()  
3  0.839652  DecisionTreeClassifier(class_weight='balanced'...  
4  0.865943  (DecisionTreeClassifier(max_features='sqrt', r...  
5  0.777779  LogisticRegression(class_weight='balanced', ra...

Метрики у моделей случайный лес и решающее дерево уменьшились, а у логистической регрессии значительно увеличилась метрика F1.

### Увеличение выборки

In [23]:
print(target_train[target_train == 0].count())
print(target_train[target_train == 1].count())

4328
1126


In [24]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_upsampled, target_upsampled = upsample(features_train, target_train, 4)
print('Размеры полученных выборок:')
print(features_upsampled.shape)
print(target_upsampled.shape)
print('Баланс классов:')
print(target_upsampled[target_upsampled == 0].count())
print(target_upsampled[target_upsampled == 1].count())

Размеры полученных выборок:
(8832, 11)
(8832,)
Баланс классов:
4328
4504


<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
Кстати, для даун/апсемплинга есть готовые решения. Чтобы в будущем не писать всегда свои собственные функции апсемплинга и даунсемплинга, можно ознакомиться с использовать <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html">SMOTE</a> для увеличения выборки и <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.under_sampling.RandomUnderSampler.html">RandomUnderSampler</a> для уменьшения.
<br>Для данного проекта - это необязательно, но можно взять на вооружение в будущем, если будет лень писать свои функции, хоть они и не сложные.
    </font>
</div>

In [25]:
max_f1 = 0
intermediate_result = []
for depth in range(1, 20):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model.fit(features_upsampled, target_upsampled)  
    predicted_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    intermediate_result.append(['Дерево', depth, f1, auc_roc])
    if f1 > max_f1:
        max_f1 = f1
        max_depth = depth
        max_auc_roc = auc_roc
result_models.append(['Дерево', 'Увеличение выборки', max_depth, max_f1, max_auc_roc, model])
intermediate_result = pd.DataFrame(data=intermediate_result, columns = ['name_models', 'depth_or_est', 'f1_score', 'auc_roc'])
print('Наилучший результат:')
print('Глубина:', max_depth)
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)
print('____________________________')
print('Промежуточные результаты:')
intermediate_result

Наилучший результат:
Глубина: 5
F1-мера:  0.5735449735449736
AUC-ROC:  0.8396523192522141
____________________________
Промежуточные результаты:


name_models  depth_or_est  f1_score   auc_roc
0       Дерево             1  0.506586  0.706652
1       Дерево             2  0.529762  0.758072
2       Дерево             3  0.548936  0.794927
3       Дерево             4  0.546119  0.817898
4       Дерево             5  0.573545  0.839652
5       Дерево             6  0.563089  0.832123
6       Дерево             7  0.541353  0.810545
7       Дерево             8  0.558140  0.790115
8       Дерево             9  0.515917  0.766402
9       Дерево            10  0.505998  0.729511
10      Дерево            11  0.515254  0.716273
11      Дерево            12  0.501818  0.693214
12      Дерево            13  0.480888  0.673058
13      Дерево            14  0.480818  0.673827
14      Дерево            15  0.468421  0.662235
15      Дерево            16  0.474035  0.664015
16      Дерево            17  0.448000  0.653274
17      Дерево            18  0.459644  0.659321
18      Дерево            19  0.460081  0.659551

In [26]:
max_f1 = 0
intermediate_result = []
for est in range(10, 200, 10):
    model = RandomForestClassifier(random_state=12345, n_estimators=est) 
    model.fit(features_upsampled, target_upsampled) 
    predicted_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    intermediate_result.append(['Случайный лес', est, f1, auc_roc])
    if f1 > max_f1:
        max_f1 = f1
        max_est = est
        max_auc_roc = auc_roc
result_models.append(['Случайный лес', 'Увеличение выборки', max_est, max_f1, max_auc_roc, model])
print('Наилучший результат:')
print('Количество деревьев:', max_est)
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)
intermediate_result = pd.DataFrame(data=intermediate_result, columns = ['name_models', 'depth_or_est', 'f1_score', 'auc_roc'])
print('____________________________')
print('Промежуточные результаты:')
intermediate_result

Наилучший результат:
Количество деревьев: 70
F1-мера:  0.6200873362445415
AUC-ROC:  0.8566774105073658
____________________________
Промежуточные результаты:


name_models  depth_or_est  f1_score   auc_roc
0   Случайный лес            10  0.578102  0.830133
1   Случайный лес            20  0.590909  0.843176
2   Случайный лес            30  0.599407  0.849670
3   Случайный лес            40  0.608824  0.852000
4   Случайный лес            50  0.609467  0.854894
5   Случайный лес            60  0.614265  0.855144
6   Случайный лес            70  0.620087  0.856677
7   Случайный лес            80  0.615160  0.857967
8   Случайный лес            90  0.615160  0.857939
9   Случайный лес           100  0.608824  0.859049
10  Случайный лес           110  0.607670  0.860019
11  Случайный лес           120  0.609720  0.859057
12  Случайный лес           130  0.618557  0.858963
13  Случайный лес           140  0.607670  0.858851
14  Случайный лес           150  0.616058  0.859938
15  Случайный лес           160  0.608824  0.859941
16  Случайный лес           170  0.602056  0.859019
17  Случайный лес           180  0.610866  0.858713
18  Случайный лес           190  0.615836  0.859741

In [27]:
model = LogisticRegression(random_state = 12345) 
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
f1 = f1_score(target_valid, predicted_valid)
probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
result_models.append(['Логистическиая регрессия', 'Увеличение выборки', None, f1, auc_roc, model])
print('Результат:')
print('F1-мера: ', f1)
print('AUC-ROC: ', auc_roc)

Результат:
F1-мера:  0.30400000000000005
AUC-ROC:  0.773663293800172


In [28]:
models = pd.DataFrame(data=result_models, columns = ['name_models', 'balance', 'depth_or_est', 'f1_score', 'auc_roc', 'models'])
models

name_models              balance  depth_or_est  f1_score  \
0                    Дерево          Без баланса           7.0  0.576433   
1             Случайный лес          Без баланса         130.0  0.591736   
2  Логистическиая регрессия          Без баланса           NaN  0.304000   
3                    Дерево  Взвешивание классов           5.0  0.573545   
4             Случайный лес  Взвешивание классов         150.0  0.581081   
5  Логистическиая регрессия  Взвешивание классов           NaN  0.509731   
6                    Дерево   Увеличение выборки           5.0  0.573545   
7             Случайный лес   Увеличение выборки          70.0  0.620087   
8  Логистическиая регрессия   Увеличение выборки           NaN  0.304000   

    auc_roc                                             models  
0  0.834605  DecisionTreeClassifier(max_depth=19, random_st...  
1  0.864223  (DecisionTreeClassifier(max_features='sqrt', r...  
2  0.773663                               LogisticRegression()  
3  0.839652  DecisionTreeClassifier(class_weight='balanced'...  
4  0.865943  (DecisionTreeClassifier(max_features='sqrt', r...  
5  0.777779  LogisticRegression(class_weight='balanced', ra...  
6  0.839652  DecisionTreeClassifier(max_depth=19, random_st...  
7  0.856677  (DecisionTreeClassifier(max_features='sqrt', r...  
8  0.773663             LogisticRegression(random_state=12345)

Метрики у моделей логистическая регрессия и решающее дерево уменьшились, а у случайного леса увеличилась метрика F1.

### Уменьшение выборки

In [29]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

features_downsampled, target_downsampled = downsample(features_train, target_train, 0.4)
print('Размеры полученных выборок:')
print(features_upsampled.shape)
print(target_upsampled.shape)
print('Баланс классов:')
print(target_upsampled[target_upsampled == 0].count())
print(target_upsampled[target_upsampled == 1].count())

Размеры полученных выборок:
(8832, 11)
(8832,)
Баланс классов:
4328
4504


In [30]:
max_f1 = 0
intermediate_result = []
for depth in range(1, 20):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model.fit(features_downsampled, target_downsampled)  
    predicted_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    intermediate_result.append(['Дерево', depth, f1, auc_roc])
    if f1 > max_f1:
        max_f1 = f1
        max_depth = depth
        max_auc_roc = auc_roc
result_models.append(['Дерево', 'Уменьшение выборки', max_depth, max_f1, max_auc_roc, model])
intermediate_result = pd.DataFrame(data=intermediate_result, columns = ['name_models', 'depth_or_est', 'f1_score', 'auc_roc'])
print('Наилучший результат:')
print('Глубина:', max_depth)
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)
print('____________________________')
print('Промежуточные результаты:')
intermediate_result

Наилучший результат:
Глубина: 5
F1-мера:  0.5853658536585367
AUC-ROC:  0.8356634152896327
____________________________
Промежуточные результаты:


name_models  depth_or_est  f1_score   auc_roc
0       Дерево             1  0.506586  0.706652
1       Дерево             2  0.528902  0.758072
2       Дерево             3  0.554455  0.797861
3       Дерево             4  0.561224  0.820354
4       Дерево             5  0.585366  0.835663
5       Дерево             6  0.580882  0.826405
6       Дерево             7  0.578512  0.810002
7       Дерево             8  0.564920  0.799206
8       Дерево             9  0.508754  0.763177
9       Дерево            10  0.496016  0.730293
10      Дерево            11  0.503375  0.728847
11      Дерево            12  0.488534  0.708521
12      Дерево            13  0.491576  0.703866
13      Дерево            14  0.488323  0.697080
14      Дерево            15  0.505241  0.707753
15      Дерево            16  0.499482  0.696714
16      Дерево            17  0.484787  0.699361
17      Дерево            18  0.481951  0.693558
18      Дерево            19  0.500000  0.705973

In [31]:
max_f1 = 0
intermediate_result = []
for est in range(10, 200, 10):
    model = RandomForestClassifier(random_state=12345, n_estimators=est) 
    model.fit(features_downsampled, target_downsampled) 
    predicted_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    intermediate_result.append(['Случайный лес', est, f1, auc_roc])
    if f1 > max_f1:
        max_f1 = f1
        max_est = est
        max_auc_roc = auc_roc
result_models.append(['Случайный лес', 'Уменьшение выборки', max_est, max_f1, max_auc_roc, model])
print('Наилучший результат:')
print('Количество деревьев:', max_est)
print('F1-мера: ', max_f1)
print('AUC-ROC: ', max_auc_roc)
intermediate_result = pd.DataFrame(data=intermediate_result, columns = ['name_models', 'depth_or_est', 'f1_score', 'auc_roc'])
print('____________________________')
print('Промежуточные результаты:')
intermediate_result

Наилучший результат:
Количество деревьев: 170
F1-мера:  0.6352657004830917
AUC-ROC:  0.8665300217024354
____________________________
Промежуточные результаты:


name_models  depth_or_est  f1_score   auc_roc
0   Случайный лес            10  0.592686  0.836608
1   Случайный лес            20  0.602941  0.842828
2   Случайный лес            30  0.613333  0.852243
3   Случайный лес            40  0.608173  0.857607
4   Случайный лес            50  0.616687  0.860717
5   Случайный лес            60  0.621359  0.863138
6   Случайный лес            70  0.626829  0.863199
7   Случайный лес            80  0.630435  0.864299
8   Случайный лес            90  0.629674  0.865106
9   Случайный лес           100  0.623942  0.865009
10  Случайный лес           110  0.632727  0.865204
11  Случайный лес           120  0.626214  0.865396
12  Случайный лес           130  0.629540  0.866314
13  Случайный лес           140  0.627262  0.866236
14  Случайный лес           150  0.621524  0.866388
15  Случайный лес           160  0.631197  0.866468
16  Случайный лес           170  0.635266  0.866530
17  Случайный лес           180  0.632850  0.866696
18  Случайный лес           190  0.628916  0.866691

In [32]:
model = LogisticRegression(random_state = 12345) 
model.fit(features_downsampled, target_downsampled) 
predicted_valid = model.predict(features_valid)
f1 = f1_score(target_valid, predicted_valid)
probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
result_models.append(['Логистическиая регрессия', 'Уменьшение выборки', None, f1, auc_roc, model])
print('Результат:')
print('F1-мера: ', f1)
print('AUC-ROC: ', auc_roc)

Результат:
F1-мера:  0.4987775061124694
AUC-ROC:  0.7713568347945816


In [33]:
models = pd.DataFrame(data=result_models, columns = ['name_models', 'balance', 'depth_or_est', 'f1_score', 'auc_roc', 'models'])
models

name_models              balance  depth_or_est  f1_score  \
0                     Дерево          Без баланса           7.0  0.576433   
1              Случайный лес          Без баланса         130.0  0.591736   
2   Логистическиая регрессия          Без баланса           NaN  0.304000   
3                     Дерево  Взвешивание классов           5.0  0.573545   
4              Случайный лес  Взвешивание классов         150.0  0.581081   
5   Логистическиая регрессия  Взвешивание классов           NaN  0.509731   
6                     Дерево   Увеличение выборки           5.0  0.573545   
7              Случайный лес   Увеличение выборки          70.0  0.620087   
8   Логистическиая регрессия   Увеличение выборки           NaN  0.304000   
9                     Дерево   Уменьшение выборки           5.0  0.585366   
10             Случайный лес   Уменьшение выборки         170.0  0.635266   
11  Логистическиая регрессия   Уменьшение выборки           NaN  0.498778   

     auc_roc                                             models  
0   0.834605  DecisionTreeClassifier(max_depth=19, random_st...  
1   0.864223  (DecisionTreeClassifier(max_features='sqrt', r...  
2   0.773663                               LogisticRegression()  
3   0.839652  DecisionTreeClassifier(class_weight='balanced'...  
4   0.865943  (DecisionTreeClassifier(max_features='sqrt', r...  
5   0.777779  LogisticRegression(class_weight='balanced', ra...  
6   0.839652  DecisionTreeClassifier(max_depth=19, random_st...  
7   0.856677  (DecisionTreeClassifier(max_features='sqrt', r...  
8   0.773663             LogisticRegression(random_state=12345)  
9   0.835663  DecisionTreeClassifier(max_depth=19, random_st...  
10  0.866530  (DecisionTreeClassifier(max_features='sqrt', r...  
11  0.771357             LogisticRegression(random_state=12345)

Метрики у модели логистическая регрессия все также показывают не очень хорошие результаты, у решающего дерева значения метрик повысились незначительно, а у случайного леса увеличились обе метрики.

### Промежуточные выводы:

1. Были применяны 3 вида борьбы с дисбалансом: взвешивание классов, увеличение выборки и уменьшение выборки.
2. При использование взвешивания классов результаты изменились в худшую сторону в моделях случайный лес и решающее дерево, результаты модели логистической регрессии значительно увеличились, ног недостаточно для достижения цели
3. При использовании увеличения выборки повысились результаты модели случайный лес, в остальных двух моделях результаты ухудшились
4. При использовании уменьшения выборки результаты немного увеличились у модели решающее дерево, модель логистической регрессии показывает наихудшие результаты среди других. Метрики у модели случайный лес увеличились и показали наилучший результат среди всех построенных моделей.
5. Наилучшей моделью принимаем модель 10: случайный лес с 170 деревьями.

## Тестирование модели

In [34]:
final_model = models.loc[10]
predicted_test = final_model['models'].predict(features_test)
final_model['f1_score'] = f1_score(target_test, predicted_test)
probabilities_test = final_model['models'].predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
final_model['auc_roc'] = roc_auc_score(target_test, probabilities_one_test)
final_model

name_models                                         Случайный лес
balance                                        Уменьшение выборки
depth_or_est                                                170.0
f1_score                                                 0.611253
auc_roc                                                  0.856572
models          (DecisionTreeClassifier(max_features='sqrt', r...
Name: 10, dtype: object

### Промежуточные выводы

На тестовый выборке получилось превысить значение в 0,59 у F1-меры и получить хороший результат метрики AUC-ROC, таким образом наилучшая модель найдена.

## Итоговые выводы:

### Ход работы:
1. Изучен набор данных, устранены пропущенные значения и удалены ненужные параметры.
2. Набор данных был разделен на три выборки: обучающая выборка, валидационная выборка и тестовая выборка в соотношении 3:1:1 соответственно
3. Проведено кодирование и масштабирование признаков.
4. Исследована модель решающее дерево с различными гипперпараметрами от 1 до 9
5. Исследована модель случайный лес с различными гипперпараметрами от 10 до 190 с шагом в 10
6. Исследована модель логистической регрессии
7. Проведена борьба с дисбалансом тремя различными способами: взвешивание классов, увеличение выборки, уменьшение выборки
8. Результаты по всем полученным моделям сведены в таблицу
9. Выбрана наилучшая модели и протестирована на тестовой выборке.

### Результаты:
1. Наилучшей моделью принимаем модель под номером 10 в сводной таблице: случайный лес с гиперпараметром равным 170
2. Результаты на тестовой выборке представлены ниже
|                         |              |
|-------------------------|--------------|
|Имя модели               |Случайный лес |
|Количество шагов в моделе|170           |
|F1-score                 |0.611253      |
|AUC-ROC                  |0.856572      |

Поставленная цель достигнута. 

<div class="alert alert-success">
<font size="4"><b>✔️ Комментарий ревьюера V2</b></font>
    <br /> 
    <font size="3", color = "black">
<br />    
Получена модель, с запасом выполняющая основное условие проекта. Кстати, здорово, что ты хорошие модели хранишь и тебе для тестирования не надо заново обучать.
<br />

<div class="alert alert-info">
Комментарий студента:

По ходу выполнения проекта возникли некоторые вопросы:
1. Не очень понятно, почему мы используем F1-меру, чем она лучше accuracy?
2. Как достичь нужной отметки, возникает только идея с изменением гипперпараметров?
3. Почему у логистической регресии в большинстве случаев F1-мера равна 0?
</div>   

<div class="alert alert-warning">
<font size="4"><b>⚠️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />

1. Про то, почему в задачах с дисбалансом метрика accuracy не рассматривается, я написал в начале работы. F1 - устойчивая к дисбалансу метрика. Как и ROC AUC

2. Во-первых, да, поиграйся с гиперпараметрами ещё. Например, кроме числа деревьев можешь попробовать для случайного леса менять и глубину дерева. Во-вторых, организуй корректно борьбу с дисбалансом, проверяй, что в блоках с ап- и даунсемплом достигается баланс. В-третьих, удали признаки, не несущие в себе смысла для нашей задачи

3. Логистическая регрессия - это линейный классификатор. То есть такой, который границу разделения классов строит прямую в пространстве признаков. Очевидно, если у неё плохой результат, значит классы плохо разделимы прямой

А теперь пришло время тебе ещё раз пытаться добиться нужного качества на тестовой выборке :)

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [X]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*